src/storage/compression/roaring.cpp

```diff
@@ -0,0 +1,1225 @@
+				}
+			}
+			count += amount;
+			return;
+		}
+
+		// Adjust the runs
+		auto &current_run_idx = run_idx[null];
+		auto &last_run_idx = run_idx[last_is_null];
+
+		if (count && null != last_is_null && last_run_idx < MAX_RUN_IDX) {
+			auto &last_run = runs[last_is_null][last_run_idx];
+			// End the last run
+			last_run.length = (count - last_run.start) - 1;
+			last_run_idx++;
+		}
+		if (!count || (null != last_is_null && current_run_idx < MAX_RUN_IDX)) {
+			auto &current_run = runs[null][current_run_idx];
+			// Initialize the new run
+			current_run.start = count;
+		}
+
+		// Add to the array
+		auto &current_array_idx = array_idx[null];
+		if (current_array_idx < MAX_ARRAY_IDX) {
+			if (current_array_idx + amount <= MAX_ARRAY_IDX) {
+				for (idx_t i = 0; i < amount; i++) {
+					arrays[null][current_array_idx + i] = count + i;
+				}
+			}
+			current_array_idx += amount;
+		}
+
+		last_is_null = null;
+		null_count += null * amount;
+		count += amount;
+	}
+
+	bool IsFull() const {
+		return count == ROARING_CONTAINER_SIZE;
+	}
+
+	void OverrideArray(data_ptr_t destination, bool nulls) {
+		arrays[nulls] = reinterpret_cast<uint16_t *>(destination);
+	}
+	void OverrideRun(data_ptr_t destination, bool nulls) {
+		runs[nulls] = reinterpret_cast<RunContainerRLEPair *>(destination);
+	}
+	void OverrideUncompressed(data_ptr_t destination) {
+		uncompressed = reinterpret_cast<validity_t *>(destination);
+	}
+
+	void Finalize() {
+		D_ASSERT(!finalized);
+		auto &last_run_idx = run_idx[last_is_null];
+		if (count && last_run_idx < MAX_RUN_IDX) {
+			auto &last_run = runs[last_is_null][last_run_idx];
+			// End the last run
+			last_run.length = (count - last_run.start);
+			last_run_idx++;
+		}
+		finalized = true;
+	}
+
+	Result GetResult() {
+		D_ASSERT(finalized);
+		const bool can_use_null_array = array_idx[NON_NULLS] < MAX_ARRAY_IDX;
+		const bool can_use_non_null_array = array_idx[NULLS] < MAX_ARRAY_IDX;
+
+		const bool can_use_null_run = run_idx[NON_NULLS] < MAX_RUN_IDX;
+		const bool can_use_non_null_run = run_idx[NULLS] < MAX_RUN_IDX;
+
+		const bool can_use_array = can_use_null_array || can_use_non_null_array;
+		const bool can_use_run = can_use_null_run || can_use_non_null_run;
+		if (!can_use_array && !can_use_run) {
+			// Can not efficiently encode at all, write it uncompressed
+			return Result::BitsetContainer(count);
+		}
+		uint16_t lowest_array_cost = MinValue<uint16_t>(array_idx[NON_NULLS], array_idx[NULLS]);
+		uint16_t lowest_run_cost = MinValue<uint16_t>(run_idx[NON_NULLS], run_idx[NULLS]) * 2;
+		uint16_t uncompressed_cost =
+		    (AlignValue<uint16_t, ValidityMask::BITS_PER_VALUE>(count) / ValidityMask::BITS_PER_VALUE) *
+		    sizeof(validity_t);
+		if (MinValue<uint16_t>(lowest_array_cost, lowest_run_cost) > uncompressed_cost) {
+			// The amount of values is too small, better off using uncompressed
+			// we can detect this at decompression because we know how many values are left
```

Comment:
This might be an edge case that still requires a test

Reply:
Actually no, that's fine, we set the most significant bit of the `cardinality` field *only* when the container is uncompressed, which doesn't require any additional context to figure out